src/main/java/de/tum/cit/aet/artemis/lecture/domain/Lecture.java

```diff
@@ -136,17 +141,113 @@ public void setAttachments(Set<Attachment> attachments) {
         this.attachments = attachments;
     }
 
+    /**
+     * Get an unmodifiable list of lecture units when the objects are initialized by Hibernate.
+     * This is important so that external code does not modify the list without updating the back-references and order.
+     *
+     * @return the lecture units
+     */
     public List<LectureUnit> getLectureUnits() {
-        return lectureUnits;
+        if (Hibernate.isInitialized(lectureUnits)) {
+            return List.copyOf(lectureUnits);
+        }
+        // when not initialized, return an empty list to avoid LazyInitializationExceptions
+        return List.of();
+    }
+
+    /**
+     * Reorder the lecture units based on the given list of ordered IDs.
+     * Makes sure to update the order after reordering.
+     *
+     * @param orderedIds the list of lecture unit IDs in the desired order
+     */
+    public void reorderLectureUnits(List<Long> orderedIds) {
+        List<LectureUnit> sorted = lectureUnits.stream().sorted(Comparator.comparing(unit -> orderedIds.indexOf(unit.getId()))).toList();
+        lectureUnits.clear();
+        lectureUnits.addAll(sorted);
+        updateLectureUnitOrder();
     }
 
+    /**
+     * Set the lecture units, replacing any existing ones.
+     * Makes sure to update back-references and order.
+     *
+     * @param lectureUnits the new list of lecture units
+     */
     public void setLectureUnits(List<LectureUnit> lectureUnits) {
-        this.lectureUnits = lectureUnits;
+        this.lectureUnits.clear();
+        if (lectureUnits != null) {
+            for (LectureUnit lectureUnit : lectureUnits) {
+                addLectureUnit(lectureUnit);        // ensures back-reference and collection management
+            }
+        }
+        updateLectureUnitOrder();
     }
```

Comment:
Not sure if we care about efficiency here, for each lecture unit we call `updateLectureUnitOrder()` already in `addLectureUnit()` - we could prevent this by passing an optional boolean flag to `addLectureUnit` (which defaults to true) and is only false in this case, so we only update the order once

Reply:
You are right, that we might do it multiple times, however it's not super easy to garantuee in all cases that the order is consistent and correct... In contrast to any (potential long Hibernate database operation), this will take nano seconds